src/Elmish/Hooks/Type.purs

```diff
@@ -1,65 +1,87 @@
 -- | withHooks do
--- |   foo /\ setFoo <- useState (HookName "Foo") ""
+-- |   foo /\ setFoo <- useState ""
 -- |   pure …
 -- | ```
--- |
--- | Wrapping it in a `WriterT (Array String)` allows us to keep track of all of
--- | the hook names so `withHooks` can log an error if there are any duplicate
--- | names. If two names are duplicated and can appear in the same spot in the
--- | DOM, React might confuse one for the other, causing unexpected
--- | side-effects.
-type Hook = WriterT (Array String) (Cont ReactElement)
+type Hook = Cont ReactElement
 
--- | Given a name and a function to create a `ComponentDef` (from a render
--- | function `a -> ReactElement`), `mkHook` creates a `Hook a` and keeps track
--- | of the name so that it can track duplicates. E.g. `useEffect` uses `mkHook`
--- | like so:
+-- | Given a `ComponentName` and a function to create a `ComponentDef` (from a
+-- | render function `a -> ReactElement`), `mkHook` creates a `Hook a`. The name
+-- | can be anything, but it’s recommended to use `genComponentName` to create a
+-- | unique name based on where the hook is called from in the stack trace.
+-- | `genComponentName` accepts a `skipFrames :: Int` argument to indicate how
+-- | many frames back it should look for the call site of the hook.
+-- |
+-- | It’s also recommended to create the name in a where clause and make your
+-- | hook a function accepting one argument. Even if your hook takes more than
+-- | one argument, you can put the rest o the arguments in a lambda in the
+-- | function body:
 -- |
 -- | ```purs
--- | useEffect :: HookName -> Aff Unit -> Hook Unit
--- | useEffect name init =
+-- | myHook x = \y z -> …
+-- |   where
+-- |     name = …
+-- |
+-- | This ensures that the number of frames to skip is predictably 2. If
+-- | defining the function differently, a different number of frames can be
+-- | passed. `genComponentNameWithTrace` can be used to help find the correct
+-- | number.
+-- |
+-- | As an example of how to use `mkHook`, `useEffect` uses it like so:
+-- |
+-- | ```purs
+-- | useEffect :: Aff Unit -> Hook Unit
+-- | useEffect init =
 -- |   mkHook name \render ->
 -- |     { init: forkVoid init
 -- |     , update: \_ msg -> absurd msg
 -- |     , view: \_ _ -> render unit
 -- |     }
+-- |   where
+-- |     name = ComponentName $ genStableUUID { skipFrames: 2 }
 -- | ```
-mkHook :: forall msg state a. HookName -> ((a -> ReactElement) -> ComponentDef msg state) -> Hook a
-mkHook (HookName name) mkDef = do
-  tell [name]
-  WriterT $ cont \render ->
-    wrapWithLocalState (ComponentName name) mkDef \args ->
-      render $ Tuple args []
+mkHook :: forall msg state a. ComponentName -> ((a -> ReactElement) -> ComponentDef msg state) -> Hook a
+mkHook name mkDef =
+  cont \render -> wrapWithLocalState name mkDef render
+
+-- | Generates a `ComponentName` to be passed to `mkHook`.
+genComponentName :: { skipFrames :: Int } -> ComponentName
+genComponentName = ComponentName <<< genStableUUID_
```

Comment:
`ComponentName` is a newtype, so you can just declare the FFI bindings to return it directly. I understand it may feel a bit unstable (e.g. what if we change the definition of `ComponentName` in the future?), so not insisting, just a thought.